### ISSUE TITLE
remark

### ISSUE THREAD

**Issue by ucm1auto** [master1auto/CCP-MODULE-PWM#1]
salame
good job, but try to write the program without using mikroc pwm library.
also try to avoid using goto instruction.

**Comment by ucm1auto** [comment]
your program is correct but you have used the mikroc functions like PWM2_Init and   PWM2_Set_Duty.
it's good, but i am hoping that you write this program by configuring the ccp module registers.
for that read carefully this link :
http://www.mikroe.com/chapters/view/16/chapter-3-pic16f887-microcontroller/
go to CCP1 IN PWM MODE section.

**Comment by ucm1auto** [comment]
did you find how you can write the program or you need help.

**Comment by ROMEYSSA** [comment]
salame 
This is the program that we used configuration the ccp module registers in duty cycle 10% Please tell us is this program correcte or no?Thank you.

**Comment by ucm1auto** [comment]
the steps are correct,but i forget to inform you that the micro controller clock equal to 8 MHz, also check this line  TRISC.RC1 = 0x00; and the T2CON value.

**Comment by DJAMEL-AUTOMATIQUE** [comment]
configuration T2CON :
TOUTPS<3:0>: Timer2 Output Postscaler Select bits
          0000
TMR2ON: Timer2 On bit
                 1 = Timer2 is on
T2CKPS<1:0>: Timer2 Clock Prescale Select bits
                1x = Prescaler is 16
T2CON=0x06
                                                                                        Thank you

**Comment by ucm1auto** [comment]
why you have set the Postscaler to 1:16?

**Comment by ROMEYSSA** [comment]
SALAME
We have changed the value ​​of T2CON is it correcte or no?
Thank you.

**Comment by ucm1auto** [comment]
you must change TRISC.RC1 = 0x00; to TRISC.RC1 = 0; because TRISC.RC1 is one bit.
i will check the values of pr2, also you must change  CCPR1 to  CCPR2L because you are using ccp2 module.

**Comment by ucm1auto** [comment]
T2CON looks correct but the pr2 (pwm period) is wrong. you may notice that the clock frequency is 8 MHZ in this project.

**Comment by ROMEYSSA** [comment]
YES we have changed the value of frequency in the programme (8MHz / 4_16_2kHz)-1 = 62,so PR2=62.

**Comment by ucm1auto** [comment]
62 is correct, now let's me check the value of CCPR2L.

**Comment by ucm1auto** [comment]
also the CCPR2L is wrong.

**Comment by ucm1auto** [comment]
use this formula: Pulse Width = (CCPR1L,DC1B1,DC1B0) \* Tosc \* TMR2 Prescale Value

**Comment by ROMEYSSA** [comment]
Duty cycle = CCPR2L/PR2+1 ,(10%=CCPR2L/63 SO CCPR2L=6)    we can use this formula ? because i dont inderstande haw i can know the value of CCPRL2 in last formula?

**Comment by ucm1auto** [comment]
The second formula is 
Duty cycle= ccpr2l/(4*(pr2+1))

**Comment by ucm1auto** [comment]
I am not sure, i will confirm tomorrow.

**Comment by ROMEYSSA** [comment]
ok thnks so the value of ccpr2l=25 for duty 10%

**Comment by ucm1auto** [comment]
yes, the value for 10% is 25, now you must complete your code for 80%.
also,try to create two function named pwm_init(), and pwm_set_duty to organize the program.

**Comment by ROMEYSSA** [comment]
salame
we completed the programme is it correct?
Thank you.

**Comment by ucm1auto** [comment]
the previous code was correct,
i asked you to create your own functions, not using the mikroc functions.

**Comment by ucm1auto** [comment]
for example:
PWM2_Init(2000); 
this function just configure the CCP2CON  and load the appropriate value in the PR2 register.
so you can write your own functions.

**Comment by ROMEYSSA** [comment]
PWM2_Init(2000); 
 CCP2CON = 0x0C; 
PR2 = 62;  
like this?

**Comment by ucm1auto** [comment]
yes, it looks correct
the same logic for pwm_set_duty
also you must change the functions names to don't conflict with those of mikroc

**Comment by ROMEYSSA** [comment]
CCP2CON = 0x0C;
                PR2 = 62;

                 T2CON = 0x06;         // Timer2 ON
                 CCPR2L = 50;
                 delay_ms(2000);
                CCPR2L =25 ;
but i have a probleme in names of function and do you can write  this function in instruction while or in outdoor the programme?

**Comment by ucm1auto** [comment]
check the code, i have write an example.

**Comment by ROMEYSSA** [comment]
salam 
we have completed the programme is it correct?

**Comment by ucm1auto** [comment]
check the new code, just correct the value of CCPR2L for 80%.

**Comment by ucm1auto** [comment]
the code looks correct, however you must test it in proteus or on the real microcontroller board.
